### PR TITLE
feat(Portal): Improve rendering

### LINKED ISSUE #636: feat(Modal): support opening/closing multiple modals
### Steps

The closeOnClickOutside and closeOnEscape props don't work as expected when opening more than one modal. The wanted behavior is that it only closed the top modal when the events are triggered. 
This also causes an issue that when both modals are open and you want to click inside the modal, all modals are closed as well. I placed an input field in the example for this issue.
1. Open a modal in a modal
2. Press ESC or click outside/inside the top modal
### Expected

It will only close the top modal.
### Result

All modals are closed.
### Testcase

http://codepen.io/anon/pen/yapmQw

### PR DESCRIPTION
**Pull request by jeffcarbs**
- Fixes bug in rendering composite react components as portal children
- Add closeOnBackgroundClick option to enable multiple portals on top of each other without closing all on click outside.
- Update modal to make use of this option (fixes #636)
- Fix portal wrapper having `undefined` as a class name if none was passed.

---

This fixes a few issues uncovered when using the Portal directly. The biggest was the inability to have a non-DOM react component as the portal child. The fix was:

``` diff
-    this.portal = ReactDOM.unstable_renderSubtreeIntoContainer(
+    ReactDOM.unstable_renderSubtreeIntoContainer(
       this,
       Children.only(children),
       this.node
     )

+    this.portal = this.node.firstElementChild
```

The issue was that `unstable_renderSubtreeIntoContainer` will return the DOM node if the element to render was a ReactDOM element (e.g. `div`) and null if the element was a regular react components. This meant for components, subsequent usage of `this.portal` would fail.

### CODE REVIEW

**Comment by jeffcarbs** [comment]
@levithomason - this is ready for review 👍

File: src/addons/Portal/Portal.js
Comment (levithomason) [inline src/addons/Portal/Portal.js]: If I'm reading corretly, the "Background" is the Dimmer when using a Modal with a Dimmer.  Though, this prop name wouldn't apply to say, a Popup, or some other component.  If so, what about using something like `closeOnRootNodeClick` instead?

**Review by levithomason — CHANGES_REQUESTED** [review]

**Comment by levithomason** [comment]
A few other issues that did not have a diff to comment on:
1. The [Dimmer Variations](http://localhost:8080/modules/modal#dimmer-variations) example for `None` does not close on click outside.
2. The [Close Config](http://localhost:8080/modules/modal#close-config) example for `No Close On Click Outside` closes on click outside.  We should probably update this to show `closeOnBackgroundClick` with a Dimmer and `closeOnDocumentClick`with no Dimmer to illustrate the differences.
3. Could we also add a Multiple Modal example?
4. Depending on the defaults we settle with, this could be a `BREAKING CHANGE: ...` PR.  If the Modal no longer behaves the same with `closeOnDocumentClick` (not closing on Dimmer click).

File: src/addons/Portal/Portal.js
Comment (jeffcarbs) [reply src/addons/Portal/Portal.js:333]: I logged both so you can see:

``` js
    console.log('children', Children.only(children))
    console.log('firstElementChild', this.node.firstElementChild)
```

```
LOG: 'children', Object{$$typeof: 60103, type: 'p', key: null, ref: null, props: Object{children: 'Hi'}, _owner: null, _store: Object{}}
LOG: 'firstElementChild', <p data-reactroot="">Hi</p>
      ✔ maintains ref to DOM node with host element
LOG: 'children', Object{$$typeof: 60103, type: function Component(props) { ... }, key: null, ref: null, props: Object{}, _owner: null, _store: Object{}}
LOG: 'firstElementChild', <p data-reactroot="">Hi</p>
      ✔ maintains ref to DOM node with React component
```

In both cases we need that actual DOM node, not a reference to the component or element.

**Comment by jeffcarbs** [comment]
> The [Dimmer Variations](http://localhost:8080/modules/modal#dimmer-variations) example for `None` does not close on click outside.

This is expected behavior (works like SUI core). I added a notification popup on that button to explain what's happening. Essentially Modals can close on dimmer click. If there's no dimmer there's nothing to click on. You can manually specify `closeOnDocumentClick` which means any outside click.

> The [Close Config](http://localhost:8080/modules/modal#close-config) example for `No Close On Click Outside` closes on click outside.  We should probably update this to show `closeOnBackgroundClick` with a Dimmer and `closeOnDocumentClick`with no Dimmer to illustrate the differences.

Fixed

> Could we also add a Multiple Modal example?

Added

> Depending on the defaults we settle with, this could be a `BREAKING CHANGE: ...` PR.  If the Modal no longer behaves the same with `closeOnDocumentClick` (not closing on Dimmer click).

No defaults changed. Closing on dimmer click still works as expected. When dimmer is set to `false` that essentially means "no dimmer" and since Modals can close on dimmer click, if there's no dimmer there's nothing to click on.

**Comment by jeffcarbs** [comment]
@levithomason - I think this should be g2g 👍 I also wound up renaming the modal examples to be inline with the new naming conventions.

**Comment by levithomason** [comment]
Cool, we can merge this, but it doesn't actually fix #636 since Escape and a click on the Dimmer still closes all modals.  It should only close the top modal, then escape again would close the next modal. I'll update the description before merging.

**EDIT**

Hm, it looks like the intent was to allow closing one modal at a time. I'll wait for your response on what you'd like to do here.  Pressing Escape/clicking the Dimmer does close both Modals.  Here, I open both Modals then click the Dimmer.  Second, I repeat but this time hit Escape.  Notice both Modals close at the same time for both cases:

**Comment by jeffcarbs** [comment]
@levithomason - The example currently functions exactly like SUI core (see http://semantic-ui.com/modules/modal.html#/examples). Clicking the dimmer or pressing escape closes all open modals.

IMHO, this opens the possibility of multiple modals being open at once, and clicking outside closing only one. The example sets dimmer to false for the second modal, but if you left the default each modal would have its own dimmer so a click on the top modal's dimmer would only close that one. Escape is handled at the document level so as it stands it would always close all.

That being said, I'd be OK merging this but not closing #636. To truly handle that, I had an idea: Create a `Modal.Group` that would take an array of `Modal`s (either via prop or children) and show them in succession within the same Portal. We could implement the `allowMultiple` prop like SUI core:
- `allowMultiple: false` (default) would show each modal one at a time
- `allowMultiple: true` would stack each new modal on top of one another

This is basically just a stateful wrapper around multiple modals. Given the current `Modal` implementation, a user could create something like this themselves so I don't think it should be something we need to do for v1. **It's also not really part of SUI.** In those examples they just daisy-chain the single modals together manually. This would really just be a SUI-react feature.

---

Sorry if that was a little scattered. The summary is that I think we should merge this since it fixes bugs (the `this.portal` bug is a blocker within in my company's app) and adds more functionality and complete any future work on separate PRs.

**Comment by jeffcarbs** [comment]
Bump 😄

**Comment by levithomason** [comment]
Thanks for the clarification, this makes sense.

**Review by levithomason — APPROVED** [review]

**Comment by levithomason** [comment]
Released in `semantic-ui-react@0.56.2`

**Pull request merged**
